File: app/src/main/java/com/example/beerhive/network/BeerServiceApiClient.java
package com.example.beerhive.network;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class BeerServiceApiClient {
    private static final String baseUrl = "https://api.punkapi.com/v2/beers";


    public static BeerServiceApi beerServiceApi;

    private BeerServiceApiClient() {

    }

    public static BeerServiceApi getInstance() {
        Retrofit retrofit;
        if (beerServiceApi == null) {
            Gson gson = new GsonBuilder()
                    .create();
            retrofit = new Retrofit.Builder()
                    .baseUrl(baseUrl)
                    .addConverterFactory(GsonConverterFactory.create(gson))
                    .build();
            beerServiceApi = retrofit.create(BeerServiceApi.class);
        }
        return beerServiceApi;
    }

}

File: app/src/main/java/com/example/beerhive/network/BeerServiceApiClient.kt
package com.example.beerhive.network

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.jakewharton.retrofit2.adapter.kotlin.coroutines.CoroutineCallAdapterFactory
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

private const val baseUrl = "https://api.punkapi.com/v2/"

private val gson: Gson = GsonBuilder().create()
private val retrofit = Retrofit.Builder()
        .addConverterFactory(GsonConverterFactory.create(gson))
        .addCallAdapterFactory(CoroutineCallAdapterFactory())
        .baseUrl(baseUrl)
        .build()

object BeerServiceApiClient {
    val beerServiceApi: BeerServiceApi by lazy { retrofit.create(BeerServiceApi::class.java) }
}
